app/services/order_cycle_distributed_products.rb

```diff
@@ -1,41 +1,21 @@
-# Finds valid products distributed by a particular distributor in an order cycle
-#
-# If a product without variants is added to an order cycle, and then some
-# variants are added to that product, but not the order cycle, then the master
-# variant should not available for customers to purchase. This class filters
-# out such products so that the customer cannot purchase them.
+# Finds products distributed by a particular distributor in an order cycle
 class OrderCycleDistributedProducts
   def initialize(order_cycle, distributor)
     @order_cycle = order_cycle
     @distributor = distributor
   end
 
-  # Returns an ActiveRecord relation without invalid products. Check
-  # #valid_products_distributed_by for details
+  # Returns an ActiveRecord relation with products distributed by the distributor in the Order Cycle
   #
   # @return [ActiveRecord::Relation<Spree::Product>]
   def relation
     variants = order_cycle.variants_distributed_by(distributor)
     products = variants.map(&:product).uniq
-
-    valid_products = products.reject do |product|
-      product_has_only_obsolete_master_in_distribution?(product, variants)
-    end
-    product_ids = valid_products.map(&:id)
-
+    product_ids = products.map(&:id)
```

Comment:
IMO next step here is to remove that 

```
variants.map(&:product)
products.map(&:id)
Spree::Product.where(id: product_ids)
```

getting the products of the variants with likely N+1 queries, then collecting their ids to then fetch again the product records from DB is a total waste of system resources.

Would you mind creating an issue?

Reply:
ah, wait, off course! Let's do that now...
We avoid testing it twice...
1 new commit, 2 less SQL queries :-)

BEFORE
- SELECT "spree_variants".* FROM "spree_variants" INNER JOIN "exchange_variants" ON "exchange_variants"."variant_id" = "spree_variants"."id" INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id" LEFT OUTER JOIN (SELECT * from inventory_items WHERE enterprise_id = 472) AS o_inventory_items ON o_inventory_items.variant_id = spree_variants.id WHERE "exchanges"."order_cycle_id" = 121 AND "exchanges"."incoming" = 'f' AND "exchanges"."receiver_id" = 472 AND ("spree_variants".deleted_at IS NULL) AND (o_inventory_items.id IS NULL OR o_inventory_items.visible = ('t'))
- SELECT "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = 291 LIMIT 1
- SELECT "spree_products".* FROM "spree_products" WHERE "spree_products"."id" IN (291) AND ("spree_products".deleted_at IS NULL)

AFTER
- SELECT "spree_products".* FROM "spree_products" WHERE "spree_products"."id" IN (SELECT spree_products.id FROM "spree_variants" INNER JOIN "exchange_variants" ON "exchange_variants"."variant_id" = "spree_variants"."id" INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id" LEFT OUTER JOIN (SELECT * from inventory_items WHERE enterprise_id = 478) AS o_inventory_items ON o_inventory_items.variant_id = spree_variants.id WHERE "exchanges"."order_cycle_id" = 124 AND "exchanges"."incoming" = 'f' AND "exchanges"."receiver_id" = 478 AND ("spree_variants".deleted_at IS NULL) AND (o_inventory_items.id IS NULL OR o_inventory_items.visible = ('t'))) AND ("spree_products".deleted_at IS NULL)

Reply:
![](https://media.giphy.com/media/xiE8iBc66lC6rEIyc3/giphy.gif)